### ISSUE TITLE
Allow MFArgs for on_mount hooks

### ISSUE THREAD

**Issue by mcrumm** [phoenixframework/phoenix_live_view#1630]
Right now we have:

```elixir
on_mount MyHook

on_mount {MyHook, :on_mount}
```

...but we should allow extra args, for example:

```elixir
on_mount {MyHook, :mount_with_args, [:foo, :bar]}
```

which would invoke MyHook.mount_with_args/5:

```elixir
def MyHook do

  # Lifecycle callback w/ additional args
  def mount_with_args(params, session, socket, :foo, :bar)
    # ...
    {:cont, socket}
  end

end
```

To do so, we need to update both the `LiveView.on_mount/1` macro and the option on `LiveView.Router.live_session/3`.

**Comment by dvic** [comment]
Related to https://github.com/phoenixframework/phoenix_live_view/issues/1629#issuecomment-921573304 I would like to discuss another alternative. I realize it's a breaking change but I'd still like to hear your thoughts on it.

Instead of allowing module/function/args, how about introducing a behavior similar to `Plug`? Has this been discussed before?

```elixir
defmodule Phoenix.Hook do
    @type opts ::
            binary
            | tuple
            | atom
            | integer
            | float
            | [opts]
            | %{optional(opts) => opts}
            | MapSet.t()

    @callback init(opts) :: opts
    
    @callback on_mount(
      unsigned_params() | :not_mounted_at_router,
      session :: map,
      socket :: Socket.t(),
      opts()
    ) :: {:cont | :halt, Socket.t()}
end
```

And then you would have `on_mount MyHook` or `on_mount {MyHook, my_opts}`.

**Comment by josevalim** [comment]
Yeah, that could be another approach. on_mount always receive 4 arguments, one of them being options. The downside of this approach is that now, for every callback, you need to define a separate module, and I find that quite wasteful. Plugs do allow both function and module plugs, which offset that, and I am not sure having both types here is warranted, as it is a much smaller surface than Plug.

**Comment by dvic** [comment]
> The downside of this approach is that now, for every callback, you need to define a separate module, and I find that quite wasteful.

I agree, you should always be able to use custom functions. 

> Plugs do allow both function and module plugs, which offset that, and I am not sure having both types here is warranted, as it is a much smaller surface than Plug.

Can you clarify "smaller surface than Plug", you mean how it's used/applied within the framework?

Ideally to me the interface would be similar to the `plug` macro, but I'm not sure this is worth the breaking change now. It can be also without the `init` callback, like this:

```elixir
# behavior
on_mount MyHook               # calls MyHook.on_mount/4 with empty opts
on_mount MyHook, my_opt: 123  # calls MyHook.on_mount/4

# custom function inside module
on_mount :foo               # calls foo/4 with empty opts
on_mount :foo, my_opt: 123  # calls foo/4

# custom function in other module
on_mount {MyModule, :foo}, my_opt: 123  # calls MyModule.foo/4
on_mount {MyModule, :foo}               # calls MyModule.foo/4 with empty opts
```

**Comment by josevalim** [comment]
> Can you clarify "smaller surface than Plug", you mean how it's used/applied within the framework?

I mean that you use way fewer on mount callbacks than plugs.

> # custom function in other module
> on_mount {MyModule, :foo}, my_opt: 123  # calls MyModule.foo/4
> on_mount {MyModule, :foo}               # calls MyModule.foo/4 with empty opts

Remember `on_mount` is also an option in `live_session`, so you need to figure out how to encode the above in `on_mount`. My take is that the above is introducing a new idiom, that is not used anywhere within Elixir and Phoenix, so I would prefer to avoid it. We either stick with `Module | {Module, opts}`, which is module based, or `Module | {Module, :function} | {Module, :function, args}`, which is function based (my vote).

**Comment by dvic** [comment]
> I mean that you use way fewer on mount callbacks than plugs.

Clear.

> Remember `on_mount` is also an option in `live_session`, so you need to figure out how to encode the above as part of the `on_mount` option. My take is that the above is introducing a new idiom, that is not used anywhere within Elixir and Phoenix, so I would prefer to avoid it. 

Hmm, yeah it does indeed look like `{M, F}, A`, which is indeed a bit weird.

> We either stick with `Module | {Module, opts}`, which is module based, or `Module | {Module, :function} | {Module, :function, args}`, which is function based (my vote).

If I would have to vote now I would go for `Module | {Module, opts}` (with or without the `init` callback) for three reasons:
- I expect most of the time to have a module anyways
- I like this pattern that uses a behavior :) i.e., there are framework specific args and one optional `opts` arg for configuring your implementation
- I think one aspect we have to keep in mind is the use-case of libraries (think of an autocomplete component that needs to install handle_info/handle_event hooks). It would be nice to be able to do `on_mount SomeLib.Hook, opt1: 123, opt2: 456` and not require `on_mount SomeLib.Hook, :on_mount_opts, [opt1: 123, opt2: 456]`, because the latter, if I'm not mistaken, will be needed in the context of a lib where options can be added in new releases.

Anyways, thanks for the discussion and taking the time to respond 👍

**Comment by chrismccord** [comment]
@dvic We still support `on_mount MyModule` so I think the only thing we are missing is `init`, which I'm not convinced we need in this case? 

@mcrumm I also think keyword list options would be preferable over var args since it mirrors a lot of what we see, such as child specs, which resolves the point that @schrockwell raised.

**Comment by mcrumm** [comment]
@chrismccord Keyword opts is less compelling without `init` imo, but I'm also not convinced we need it.

To José's point, we introduce a new paradigm if we allow `{Module, :function, Keyword.t}` which doesn't exist anywhere I can think of. But if we remove the ability to specify a function name, then it forces the user to define a whole module per lifecycle on_mount hook. Given those two concepts are at odds we either have to pick one or introduce the third thing :)

**Comment by dvic** [comment]
> We still support `on_mount MyModule` so I think the only thing we are missing is `init`, which I'm not convinced we need in this case?

The `init` is not that important, I agree. 

> I also think keyword list options would be preferable over var args since it mirrors a lot of what we see, such as child specs, which resolves the point that @schrockwell raised.

To me this also feels more natural, especially considering the library use-case I mentioned above.

**Comment by chrismccord** [comment]
@mcrumm to be clear, we would only support {M, F, Keyword.t} so there is no single callback per module. We could remap `{M, :f}` to `{M, :f, []}` as well, or rely on the user to `opts \\ []` it.

**Comment by chrismccord** [comment]
errr, we'd support the following:

    on_mount MyModule
    on_mount {MyModule, :func}
    on_mount {MyModule, :func, foo: :bar}

**Comment by dvic** [comment]
> But if we remove the ability to specify a function name, then it forces the user to define a whole module per lifecycle on_mount hook

@mcrumm I wanted to come back on this point José made. I just realized that this isn't really a problem anymore once you allow a opts/config to be passed right? If I'm not mistaken, you can easily do

```elixir
on_mount MyHooks, :hook1
on_mount MyHooks, :hook2
on_mount MyHooks, {:hook3, with_some_opt: :foo}
```

Then `MyHooks` can be

```elixir
defmodule MyHooks do

  @impl Phoenix.LiveView.Hook
  def on_mount(params, session, socket, :hook1) do
  end

  @impl Phoenix.LiveView.Hook
  def on_mount(params, session, socket, :hook2) do
  end

  # etc.
end
```

**Comment by dvic** [comment]
By the way, here I assume that `on_mount MyHook` results in `MyHook. on_mount/4` being called (without `nil` as `opts` or whatever `init` returned if we'd have that, that could be an optional callback).

**Comment by chrismccord** [comment]
We need it if you want to pass any thing to the callback, such as a `role` for your authentication code. Var args could of course work, but I think the syntax will be unusual compared to the tuple form folks are used to when wanting to pass keyword options, which are pretty ubiquitous. For example the following could would be allowed under the current proposal:

    on_mount {UserAuth, :authenticate_mount, role: :admin, key: :user_id}

But it would be invoked as `UserAuth.authenticate_mount(params, session, socket, {role: :admin}, {key: :user_id}))` which I think would be surprising to many folks.

**Comment by dvic** [comment]
Yes, I also think that most users will find that surprising, but library authors might also not expect that they have to implement both `on_mount/3` (no options) and `on_mount/4` (with options). I guess you can do that easily with `def on_mount(params, session, socket, opts \\ [])` but won't this cause confusion? Isn't it better to just always have an `opts` arguments just like `Plug.call/2` has?

Given the fact that then you can do things like `on_mount MyHooks, :hook1` I'm starting to think that we might not need the `{Module, func, arg}` at all? What is then the added benefit of supporting this additional syntax? We can then have a well-defined behaviour `Phoenix.LiveView.Hook` and it will work exactly the same way `Plug.call/2` works (we can omit `init/1`).

**Comment by chrismccord** [comment]
Further thinking about this, I think @mcrumm proposal is the most straight forward option. I think the `{M, F, Keyword.t()}` will absolutely trip folks up with match errors, I also think we should avoid `{M, Keyword.t()}` in favor of multiple functions, so MFA is the most straight forward on what module + function + args means to folks. We could even raise if we detect a keyword and instruct what the user needs to change. We could also rewrite the keyword and wrap as a single list argument, but I think special casing also can be confusing.

**Comment by dvic** [comment]
@chrismccord and what about libraries? Let's say I write some hook library `SomeLibHook`. Then users will have to call it (in practice) like 

`on_mount {SomeLibHook, :on_mount, [role: :admin, key: :user_id]}`, 

correct? 

To me, 

`on_mount SomeLibHook, role: :admin, key: :user_id` 

just feels more natural. Note that this is `mount/2` and not `mount/1`, I agree that `mount/1` with `{M, Keyword.t()}` is confusing.

**Comment by mcrumm** [comment]
@dvic With your `mount/2` example, one must specify a separate mount function per module.

One case we've been exploring is on the `phx.gen.auth` generator, where we will definitely have more than one mount function:

```elixir
def MyAppWeb.LiveAuth do
  def mount_current_user(params, session, socket), do: # ...

  def require_mounted_user(params, session, socket), do: # ...
end
```

Would it be better to give each of these their own module? I am not convinced.

**Comment by dvic** [comment]
> With your mount/2 example, one must specify a separate mount function per module.

I don't agree with that, see below 🙂 

> Would it be better to give each of these their own module? I am not convinced.

@mcrumm I agree, these belong to the same module but for me the solution I proposed above would be sufficient, i.e., `on_mount MyAppWeb.LiveAuth, :mount_current_user` and then

```elixir
def MyAppWeb.LiveAuth do
  def on_mount(params, session, socket, :mount_current_user), do: # ...

  def on_mount(params, session, socket, :require_mounted_user), do: # ...
end
```

Having to do this might not be perfect but neither is having to do `on_mount {SomeLibHook, :on_mount, [role: :admin, key: :user_id]}`, right? So like you said, it boils down to picking one or adding a third option 🙂

I can live with your approach (so just having `mount/1`), but I'd love to have `mount/2` that works with an explicit behavior with a `on_mount/4`, but José already indicated that he would prefer not to support both options. We could of course just go with this PR and then wait and see if there is a solid need for `mount/2` (maybe I'm putting too much weight on the whole library thing).

**Comment by mcrumm** [comment]
@dvic Ah, I see now what you meant. This is why we did not add args at the beginning– there are a number of considerations to make :) I appreciate your contributions to the discussion!

**Comment by josevalim** [comment]
I am rethinking this and I think we should go with `on_mount M` and `on_mount {M, arg}` and always call `on_mount/4`. It is the cleanest API and in the docs we will say this:

```
Tip: if you need to define multiple `on_mount` callbacks, avoid
defining multiple modules. Instead, a tuple and pattern matching
to handle different cases:

    def on_mount(:admin, _, _, _) do
    end

    def on_mount(:user, _, _, _) do
    end

And then invoke it as:

    on_mount {M, :admin}
    on_mount {M, :user}

```

This is not really different from defining multiple functions. We keep the API clean and concise.

**Comment by dvic** [comment]
> I am rethinking this and I think we should go with `on_mount M` and `on_mount {M, arg}` and always call `on_mount/4`. It is the cleanest API and in the docs we will say this:
> 
> ```
> Tip: if you need to define multiple `on_mount` callbacks, avoid
> defining multiple modules. Instead, pass a tuple and use pattern
> matching to handle different cases:
> 
>     def on_mount(:admin, _, _, _) do
>     end
> 
>     def on_mount(:user, _, _, _) do
>     end
> 
> And then invoke it as:
> 
>     on_mount {M, :admin}
>     on_mount {M, :user}
> ```
> 
> This is not really different from defining multiple functions. We keep the API clean and concise.

@josevalim I'm also completely happy to go with this approach. Two questions:
- Out of curiosity, what is the reason to put the option arg first? I guess for the example usecase it's more readable but as a general signature `def on_mount(opts, params, session, socket)` it might be surprising to developers (because `opts` are usually the last argument). 
- What would you use as value for `arg` when you call `on_mount M`? Just `nil`? It's either choosing a default value or still going with an optional compile-time `init/1` callback. I don't have a specific use-case for this, and I'm also not aware of the implications if we want to support this, it would be just that we don't have to choose a default value 🙂

**Comment by josevalim** [comment]
I think we can call it `arg`, as in `on_mount(arg, params, session, socket)`, since it is not really optional.

**Comment by dvic** [comment]
> I think we can call it `arg`, as in `on_mount(arg, params, session, socket)`, 

That makes sense, then it's semantically different from the usual `opts`, good idea 👍

> since it is not really optional.

I assume you mean 'not optional' in the sense that we now always have `on_mount/4`?

**Comment by chrismccord** [comment]
@josevalim sounds great! Let's ship it!
